### DISCUSSION TITLE
How does BlazorWebView handle OnBackButtonPressed?

### DISCUSSION THREAD

**Discussion by nextfool** [dotnet/maui#5771] [Q&A] [answered]
Hi Community,

I have a question regarding how BlazorWebView can navigate to correct pages when we click the back button (or swipe the screen in case of gesture).

So in a simple Maui Blazor App project, lets say that we have some simple razor pages like bellow.
![image](https://user-images.githubusercontent.com/2167369/161377587-3d9e35ed-26ad-430c-9503-e5f68acdc80c.png)

Each page has simple markup that links them together 
![image](https://user-images.githubusercontent.com/2167369/161377975-a345f756-9732-4397-a55b-e337c2cecfa5.png)
Then I started the app, and clicked the link from Page1 all the way down to Page4, when I hit the back button, the pages are navigated all way up from Page4 to Page1, and it just works!

So my question is that, without any overriding the OnBackButtonPressed method in MainPage.cs, how does all this work as expected? 
Where does blazor handle the back button pressed event? It would be great if anyone can show me some code or give me some explanation. 

Thank you very much.

**Comment by nextfool** [comment]
Ok, I just want to add more questions to this session:
![image](https://user-images.githubusercontent.com/2167369/161390882-1d7ff1c3-235b-4867-b4f9-dd1cc414e1bc.png)
So now I added a breakpoint to the main page as the above shows, when I hit the go back button to navigate from Page4 to Page1, this breakpoint was NOT reached at all, it was only reached when page1 is shown, and when I stepped over the breakpoint, the test app exits. 

So this means that, BlazorWebView may have some kinds of its own navigation stack.
![image](https://user-images.githubusercontent.com/2167369/161391066-067a854e-89eb-4529-8e8c-64ee1d117e03.png)
I also found the above code to prove my guess, but I am not quite sure since I didn't find any thing related to the ModelStack in BlazorWebView code. 

Can anyone please help me out here? So basically I want to know how BlazorWebView get notified by the back button clicked event?

Thank you very much.

@Eilon Can you take some time to help me understand a little bit pleases:)

**Reply by Eilon** [reply]
@TanayParikh - I think you did some of the Back Button work in BlazorWebView. Maybe you can help?

**Reply by TanayParikh** [reply]
> So my question is that, without any overriding the OnBackButtonPressed method in MainPage.cs, how does all this work as expected?
> Where does blazor handle the back button pressed event? It would be great if anyone can show me some code or give me some explanation.

For Android, this is handled via:

https://github.com/dotnet/maui/pull/4364

Do we support "back" on any other platforms at the moment?

**Reply by nextfool** [reply]
@Eilon  @TanayParikh Thank you very much gentlemen, this is a huge help for me. Thank you again!

**Reply by Alby-o** [reply]
@TanayParikh - How do we override the default behaviour of the back button? I'm running into an issue where I want to close modal / popup windows in my razor components when the back button is hit, but instead the page is being navigated back

**Comment by nextfool** [comment, marked as answer]
Since the two gentlemen above only replied to the session, I'll just create a new answer here:

So for any here who wants to know the answer, [here ](https://github.com/dotnet/maui/pull/4364#discussion_r794133467)it is.

**Comment by Ricardoaraujo00** [comment]
Hello

Sorry for re-openning this thread but i'm very confused with the explanation because I don't have a clue where to use the answer code. I would like to close a modal when a android user(or IOS) clicks the back button on their device. Is there a clearer explanation on how it is done?

**Reply by Ricardoaraujo00** [reply]
It seems it can be done this simple way. Works with browser back button on PC so it should work on android also.

Inject JsRuntime on modal component
`@inject IJSRuntime JsRuntime`
 
Call Javascript function to register the modal opening as a navigation
```
protected override async Task OnInitializedAsync()
    {
        await JsRuntime.InvokeAsync<string>("AddModalToHistory");
    }

```

The javascript function
```
<script>
function AddModalToHistory(url) {
            window.history.pushState("MyApp.Client", "Modal", url);
        }
</script>
```

I'm not passing the url but it only works this way I dont know why.
Copied from here:
[https://forum.radzen.com/t/dialogservice-on-back-pressed/13171](url)
